bofire/strategies/predictives/acqf_optimization.py

```diff
@@ -0,0 +1,722 @@
+import copy
+from abc import ABC, abstractmethod
+from typing import Callable, Dict, List, Optional, Tuple, Type
+
+import pandas as pd
+import torch
+from botorch.acquisition.acquisition import AcquisitionFunction
+from botorch.optim.initializers import gen_batch_initial_conditions
+from botorch.optim.optimize import (
+    optimize_acqf,
+    optimize_acqf_discrete,
+    optimize_acqf_list,
+    optimize_acqf_mixed,
+)
+from torch import Tensor
+
+from bofire.data_models.constraints.api import (
+    LinearEqualityConstraint,
+    LinearInequalityConstraint,
+    NChooseKConstraint,
+    ProductConstraint,
+)
+from bofire.data_models.domain.api import Domain
+from bofire.data_models.enum import CategoricalEncodingEnum, CategoricalMethodEnum
+from bofire.data_models.features.api import (
+    CategoricalDescriptorInput,
+    CategoricalInput,
+    CategoricalMolecularInput,
+    DiscreteInput,
+    Input,
+)
+from bofire.data_models.molfeatures.api import MolFeatures
+from bofire.data_models.strategies.api import (
+    AcquisitionOptimizer as AcquisitionOptimizerDataModel,
+)
+from bofire.data_models.strategies.api import (
+    BotorchOptimizer as BotorchOptimizerDataModel,
+)
+from bofire.data_models.strategies.api import RandomStrategy as RandomStrategyDataModel
+from bofire.data_models.strategies.api import (
+    ShortestPathStrategy as ShortestPathStrategyDataModel,
+)
+from bofire.data_models.strategies.shortest_path import has_local_search_region
+from bofire.data_models.types import InputTransformSpecs
+from bofire.strategies.random import RandomStrategy
+from bofire.strategies.shortest_path import ShortestPathStrategy
+from bofire.utils.torch_tools import (
+    get_initial_conditions_generator,
+    get_interpoint_constraints,
+    get_linear_constraints,
+    get_nonlinear_constraints,
+    tkwargs,
+)
+
+
+class AcquisitionOptimizer(ABC):
```

Comment:
Usually, I prefer ABCs to be purely abstract. And composition over inheritance for code re-use. We can have a discussion about this at the next hackathon or over coffee some time. Since BoFire is build like this at many places we are at least consistent.